src/snowflake/connector/options.py

```diff
@@ -2,53 +2,103 @@
+
+    def __getattr__(self, item):
+        raise MissingDependencyError(self._dep_name)
+
+
+class MissingPandas(MissingOptionalDependency):
+    """The class is specifically for pandas optional dependency."""
+    _dep_name = "pandas"
+
+
+class MissingKeyring(MissingOptionalDependency):
+    """The class is specifically for sso optional dependency."""
+    _dep_name = "keyring"
+
+
+ModuleLikeObject = Union[ModuleType, MissingOptionalDependency]
 
 
 def warn_incompatible_dep(dep_name: str,
                           installed_ver: str,
                           expected_ver: 'pkg_resources.Requirement') -> None:
-    warnings.warn(
-        "You have an incompatible version of '{}' installed, please install a version that "
-        "adheres to: '{}'".format(dep_name,
-                                  _expected_version),
-        stacklevel=2)
+    warnings.warn("You have an incompatible version of '{}' installed ({}), please install a version that "
+                  "adheres to: '{}'".format(dep_name, installed_ver, expected_ver),
+                  stacklevel=2)
 
 
-class MissingPandas(object):
+def _import_or_missing_pandas_option() -> Tuple[ModuleLikeObject, ModuleLikeObject, bool]:
+    """This function tries importing the following packages: pandas, pyarrow.
 
-    def __getattr__(self, item):
-        raise MissingDependencyError('pandas')
-
-
-try:
-    import pandas
-    # since we enable relative imports without dots this import gives us an issues when ran from test directory
-    from pandas import DataFrame  # NOQA
-    import pyarrow
-
-    installed_pandas = True
-    # Make sure we have the right pyarrow installed
-    _pandas_extras = pkg_resources.working_set.by_key['snowflake-connector-python']._dep_map['pandas']
-    _expected_version = [dep for dep in _pandas_extras if dep.name == 'pyarrow'][0]
-    _installed_pyarrow = pkg_resources.working_set.by_key['pyarrow']
-    if _installed_pyarrow and _installed_pyarrow.version not in _expected_version:
-        warn_incompatible_dep('pyarrow', _installed_pyarrow.version, _expected_version)
-except ImportError:
-    pandas = MissingPandas()
-    pyarrow = MissingPandas()
-
-try:
-    import keyring
-
-    installed_keyring = True
-except ImportError:
-    keyring = None
+    If available it returns pandas and pyarrow packages with a flag of whether they were imported.
+    It also warns users if they have an unsupported pyarrow version installed if possible.
+    """
+    try:
+        pandas = importlib.import_module('pandas')  # NOQA
+        # since we enable relative imports without dots this import gives us an issues when ran from test directory
+        from pandas import DataFrame  # NOQA
+        import pyarrow  # NOQA
+        pyarrow = importlib.import_module('pyarrow')  # NOQA
+        # Check whether we have the currently supported pyarrow installed
+        installed_packages = pkg_resources.working_set.by_key
+        print(installed_packages)
```

Comment:
Is this intentional?

Reply:
or a debug message?

Reply:
Oops, I'll look over the PR one more time. Sorry!